Add achievements and leaderboard cookbook

filiph opened the pull request:
This PR adds a recipe for the `games_services` package.

Closes https://github.com/flutter/website/issues/9670

sfshaza2 commented on src/cookbook/plugins/achievements-leaderboard.md:
```diff
@@ -0,0 +1,350 @@
+
+Gamers have various motivations for playing games. In broad strokes,
+there are four major motivations: 
+[immersion, achievement, cooperation and competition](https://meditations.metavert.io/p/game-player-motivations).
```
```suggestion
[immersion, achievement, cooperation, and competition](https://meditations.metavert.io/p/game-player-motivations).
```

sfshaza2 commented on src/cookbook/plugins/achievements-leaderboard.md:
```diff
@@ -0,0 +1,350 @@
+No matter the game you build, some players want to *achieve* in it. This
+could be trophies won or secrets unlocked. Some players want to
+*compete* in it. This could be hitting high scores or accomplishing
+speedruns. These two map to the concepts of *achievements* and
```
```suggestion
speedruns. These two ideas map to the concepts of *achievements* and
```

sfshaza2 commented on src/cookbook/plugins/achievements-leaderboard.md:
```diff
@@ -0,0 +1,350 @@
+Ecosystems such as the App Store and Google Play provide centralized
+services for achievements and leaderboards. Players have achievements
+from all the games in one place and developers don't need to
+re-implement the wheel for every game.
```
```suggestion
re-implement them for every game.
```

sfshaza2 commented on src/cookbook/plugins/achievements-leaderboard.md:
```diff
@@ -0,0 +1,350 @@
+To enable Game Center (GameKit) on iOS:
+
+1.  Open your Flutter project in Xcode.
+    open `ios/Runner.xcworkspace`
```
```suggestion
    Open `ios/Runner.xcworkspace`
```

sfshaza2 commented on src/cookbook/plugins/achievements-leaderboard.md:
```diff
@@ -0,0 +1,350 @@
+in the previous example. The rest of this recipe omits exception
+handling for clarity.
+
+**Note:** It's a good practice to create a controller. This would be a
+`ChangeNotifier`, a bloc, or some other piece of logic that wraps around
+the raw functionality of the `games_services` plugin.
```
```suggestion
{{site.alert.note}}
  It's a good practice to create a controller. This would be a
  `ChangeNotifier`, a bloc, or some other piece of logic that wraps around
  the raw functionality of the `games_services` plugin.
{{site.alert.end}}
```

sfshaza2 reviewed (approved):
Very minor nits. Otherwise lgtm

sfshaza2 commented on src/cookbook/index.md:
```diff
@@ -103,6 +103,7 @@ reference to help you build up an application.
 
 
 ## Plugins
+- [Add achievements and leaderboards to your mobile game]({{site.url}}/cookbook/plugins/achievements-leaderboard)
```
@filiph, I thought you were going to add a new "## Games" section rather than hide this in "Plugins"? I'd like to see it in Games.

filiph replied:
Will do!

filiph commented:
Alright, suggestions implemented and the recipe is now under the new `cookbook/games` section.

Here's the page, hosted: https://flutter-website-filiph-staging.web.app/cookbook/games/achievements-leaderboard

I'm removing the "WIP" from the PR.

filiph commented:
Update: I also removed the placeholder redirect from `/cookbook/games/achievements-leaderboard` to `flutter.dev/games`.

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+
+5.  Close Xcode.
+
+6.  If you haven't already, register your game 
```
I'm not super familiar with iOS development or deployment, but does this require being part of the Apple developer program?

If so, can we add a mention here or the the beginning of the iOS section that this a prerequisite and link to the relevant page. I believe we have some examples of this on the site already.

filiph replied:
I feel that might be going too far? Not even the page you're mentioning (https://docs.flutter.dev/deployment/ios) talks about the Apple developer program sign up process. As much as I like to inline knowledge so people don't need to go hunting for it elsewhere, I think that teaching people here that, for an iOS game, they need an Apple Developer account, is excessive.

That said, I don't _mind_ putting it here. In the end, it's a judgement call. I'd lean towards not mentioning such things, that's all.

parlough replied:
That makes sense. I'm fine leaving the note out. Thanks!

parlough commented on src/cookbook/games/achievements-leaderboard.md:
````diff
@@ -0,0 +1,352 @@
+
+1.  Add a dependency on the [`games_services` package](https://pub.dev/packages/games_services).
+
+    ```shell
+    flutter pub add games_services
+    ```
````
To be consistent with elsewhere on the site:

````suggestion
    ```terminal
    $ flutter pub add games_services
    ```
````

filiph replied:
Fixed

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+## 2. Sign in to the game service
+
+Now that you have set up *Game Center* and *Play Games Services*, and
+have your achievement & leaderboard IDs ready, it's finally Dart time.
```
I love Dart time!

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+Now that you have set up *Game Center* and *Play Games Services*, and
+have your achievement & leaderboard IDs ready, it's finally Dart time.
+
+1.  Add a dependency on the [`games_services` package](https://pub.dev/packages/games_services).
```
```suggestion
1.  Add a dependency on the [`games_services` package]({{site.pub-pkg}}/games_services).
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+in the previous example. The rest of this recipe omits exception
+handling for clarity.
+
+{{site.alert.note}}
```
The tip aside might draw more attention to this :) 

```suggestion
{{site.alert.tip}}
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+    This displays the platform achievements UI as an overlay on your game.
+
+3.  To display the achievements in your own UI, use
+    `GamesServices.loadAchievements()`.
```
Consider linking this to the package's API docs:

```suggestion
    [`GamesServices.loadAchievements()`][].
    
[`GamesServices.loadAchievements()`]: {{site.pub-api}}/games_services/latest/games_services/GamesServices/loadAchievements.html
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
````diff
@@ -0,0 +1,352 @@
+    ```
+
+3.  If you want to display the leaderboard scores in your own UI, you
+    can fetch them with `GamesServices.loadLeaderboardScores()`.
````
Consider linking this to the package's API docs:

```suggestion
    can fetch them with [`GamesServices.loadLeaderboardScores()`][].
    
[`GamesServices.loadLeaderboardScores()`]: {{site.pub-api}}/games_services/latest/games_services/GamesServices/loadLeaderboardScores.html
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+
+There's more to the `games_services` plugin. With this plugin, you can:
+
+-   get the player's icon, name or unique ID
+
+-   save and load game states
+
+-   sign out of the game service
```
```suggestion
- Get the player's icon, name or unique ID
- Save and load game states
- Sign out of the game service
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+3.  If you want to display the leaderboard scores in your own UI, you
+    can fetch them with `GamesServices.loadLeaderboardScores()`.
+
+## 5. Complete example
```
I'm not sure "Complete example" makes sense as the title of this step, perhaps something such as `5. Next steps` or similar?

filiph replied:
Good point, fixed

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+
+Each game has different needs from game services.
+
+To start, you might want to create this controller:
```
Can you expand a bit more on "why" you might want to create this controller? Is it to encapsulate the previous steps?

filiph replied:
Ack

parlough commented on src/cookbook/games/achievements-leaderboard.md:
````diff
@@ -0,0 +1,352 @@
+
+To start, you might want to create this controller:
+
+```dart
````
I love this fuller example!

Could you update this to use code excerpts and place the source in [`/examples/cookbook/games/achievements-leaderboard`](https://github.com/flutter/website/tree/main/examples/cookbook), or create an issue so I or someone else can follow up to do so? That way we can keep up to date with changes in the packages and make sure analysis passes with Dart changes.

filiph replied:
Yes, I'll create an issue. I'll probably pick up the issue myself, in a week or so. I just want to make this ready as soon as possible.

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+-   sign out of the game service
+
+Some achievements can be incremental. For example: "You have collected
+all 10 pieces of the McGuffin.".
```
```suggestion
all 10 pieces of the McGuffin."
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+Gamers have various motivations for playing games. In broad strokes,
+there are four major motivations: 
+[immersion, achievement, cooperation, and competition](https://meditations.metavert.io/p/game-player-motivations).
+No matter the game you build, some players want to *achieve* in it. This
+could be trophies won or secrets unlocked. Some players want to
+*compete* in it. This could be hitting high scores or accomplishing
+speedruns. These two ideas map to the concepts of *achievements* and
+*leaderboards*.
```
No textual changes here, just breaking up the sentences so any future edits/diffs are easier to review. 

```suggestion
No matter the game you build, some players want to *achieve* in it.
This could be trophies won or secrets unlocked.
Some players want to *compete* in it.
This could be hitting high scores or accomplishing speedruns.
These two ideas map to the concepts of *achievements* and *leaderboards*.
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+
+![A simple graphic representing the four types of motivation explained above](/assets/images/docs/cookbook/types-of-gamer-motivations.png)
+
+Ecosystems such as the App Store and Google Play provide centralized
+services for achievements and leaderboards. Players have achievements
+from all the games in one place and developers don't need to
+re-implement them for every game.
```
When reading, I wasn't sure what "all the games" meant. Would something like "players **can view** achievements from **all their** games..." make sense?

```suggestion
Ecosystems such as the App Store and Google Play provide
centralized services for achievements and leaderboards.
Players can view achievements from all their games in one place and
developers don't need to re-implement them for every game.
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+
+## 1. Enable platform services
+
+To enable games services, set up *Game Center* on iOS and *Google Play
+Games Services* on Android.
```
Semantic breaks to make future diffs/changes easier to review.

```suggestion
To enable games services, set up *Game Center* on iOS and
*Google Play Games Services* on Android.
```

parlough commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+To enable games services, set up *Game Center* on iOS and *Google Play
+Games Services* on Android.
+
+### iOS
```
From a brief glance at the [GameKit docs](https://developer.apple.com/documentation/gamekit/), it seems it supports macOS. Would these instructions work for macOS and the Mac App Store too?

If so, might be worth calling this section "iOS and macOS" and updating the various mentions, or at least calling out somewhere in the section that these instruction can be applied to macOS too.

filiph replied:
I haven't tested this, so I'd rather underpromise than overpromise. I _think_ GameKit behaves a bit differently on macOS than on iOS, so it needs further investigation. I'm making a note to self.

parlough replied:
Sounds good, expanding things later is great too :)

parlough reviewed (approved):
This is really exciting @filiph and reads really well! Thanks so much, and sorry about the delay in reviewing.

I have a few questions, suggestions, and minor formatting/copy fixes.

If you or someone else would like to get this landed immediately and handle my questions as a followup, that's completely cool.

@atsansone I noticed you changed the branch to `games-v2`. Is there a particular reason? Can't we just land this into `main` once ready?

filiph commented on src/cookbook/games/achievements-leaderboard.md:
```diff
@@ -0,0 +1,352 @@
+from all the games in one place and developers don't need to
+re-implement them for every game.
+
+This recipe demonstrates how to use the `games_services` package to add
```
Fixed

filiph commented:
Thanks for the feedback. I think I addressed all issues, except for 2, which I'm deferring to later (issues forthcoming).

FWIW, I'm happy with the current state. Feel free to land when ready, @atsansone.

zoeyfan commented:
FYI. The URL link for this cookbook will be: 
https://docs.flutter.dev/cookbook/games/achievements-leaderboard
